### ISSUE TITLE
Discussion: UA: window vs iframe

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [pyllyukko/user.js#193]
see: https://github.com/gorhill/uMatrix/issues/649

Personally, I do not try to hide my browser as being FF, but I do always spoof it as the latest ESR. As crssi says in the above link, if you go to https://www.browserleaks.com/javascript and check your UA values in both `window` and `iframe.contentWindow` , if you are spoofing, you will get different results.
- crssi uses uMatrix to spoof
- I use UAControl and User-Agent JS-Fixer

Does anyone else get this behavior with other extensions? Is this a bug that someone can let Mozilla know about?

Edit: Assuming the user spoofs the same value for both headers and JS. The iframe leaks the real value.

**Comment by crssi** [comment]
Not only different.
Under `iframe.contentWindow` you get the real browser you are using.

**Comment by Roman-Nopantski** [comment]
@crssi : yup, I am assuming the user is spoofing the same value and covering headers and JS - cogito ergo sum, one is the real value

**Comment by BananaMangoFestival** [comment]
The latest version of Random Agent Spoofer now spoofs also the iframe, you have to enable script injection

**Comment by Roman-Nopantski** [comment]
I mentioned this elsewhere, so FYI, I personally recommend that UA spoofing is a waste of time (the determined, google for instance, if they really wanted too, could easily catch you out, and your entropy with the two values would skyrocket.

> Besides that hole in iframes, rather than navigator.userAgent or User-Agent HTTP header, you can identify a browser by:
- property order of special browser objects, such as navigator and screen objects, which depends on a browser, its version and operating system
- presence of some unique methods and properties of special browser objects, such as navigator.mozSms( Firefox ), navigator.webkitStartActivity( Chrome ), navigator.appMinorVersion( IE )
- modification of special objects with Javascript (eg checking if a script can delete an object property, replace one property with another, or delete the whole object)
- new browser features introduced in each version.

Anyone who thinks spoofing their UA is doing ANY good is deluded :)

**Comment by Roman-Nopantski** [comment]
FYI: Actually .. there are a bunch of tests here: http://www.darkwavetech.com/fingerprint/fingerprint_code.html They look pretty standard, I don't think it's worth adding to the list, but the one for "True Browser Fingerprinting" ( http://www.darkwavetech.com/fingerprint/fingerprint_truebrowser.html ) attempts to detect based on features, not the UA. I haven't tested it, and it doesn't go as far as to try to verify any version. Feel free to spoof your UA (JS & HTTP header) as chrome or something and tell me how it went for ya :)

**Comment by nodiscc** [comment]
I have just run the tests mentioned above (mostly http://www.darkwavetech.com/fingerprint/fingerprint_truebrowser.html, http://www.darkwavetech.com/fingerprint/fingerprint_os.html), using various spoofed User-Agent strings (real browser/version/OS is Firefox 53 on Linux 64 bits):

 * Mozilla/5.0 (Windows NT 6.1; rv:31.0) Gecko/20100101 Firefox/31.0 (commented-out value in user.js, same value as Tor browser)
 * Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36 (random Google Chrome UA string)

Note that javascript **has** to be enabled for the test to run.

Results:

 * Spoofed Firefox UA string: Firefox on Windows 7|32 bits
 * Spoofed Chrome UA: Firefox on Windows 7|32 bits

While fingerprinting tests can successfully uncloak a browser pretending to be a _different_ browser, there is no evidence that it can detect 1. a spoofed operating system substring 2. A spoofed version substring.

Buty anyway I suggest that we leave the User-Agent pref commented out. If you need to defend against advanced fingerprinting techniques there is a chance that you should be using Tor browser directly in the first place (UA string only adds a small amount of entropy). Spoofing UA also has indirect disadvantages (websites nagging you about your browser being too old/unsupported, disabled functionality such as on AMO, etc.)

I think that this issue can be closed.

**Comment by Thorin-Oakenpants** [comment]
"A spoofed version substring" - it can, by using `general.buildID.override` (which is covered in the js), so that that needs to be made clear. It also can by using feature detect. At the moment it's probably not even worth the bother for web-analytics companies to do this (they have easier ways, more cost effective, to grab 75-80% of the "market"), but the science behind it is absolutely 100%. That darkwavetech site was just an interesting example I came across, but is not comprehensive IMO (and one of the other scripts failed miserably, so it didn't exactly instill any confidence).

"A spoofed operating system substring" - doesn't `navigator.platform` and `navigator.oscpu` leak in JS? Correct me if this is not the case. Also (see point above),`navigator.buildID` will reveal the build down to the second and from that it can infer your OS sometimes (not a substring though? although 32 vs 64 bit builds are a thing). This may have changed, but AFAIK, that buildID is still unique per build - for example mibe is `20170302120751` for Windows 64bit 52.0. Anyone have there's to share such as same version on Linux?

"UA string only adds a small amount of entropy" - The UA string can provide a *large* amount of entropy. Panoptoclick calculates `Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:52.0) Gecko/20100101 Firefox/52.0` as over 8 bits (remember they want 32? bits theoretically to absolutely make you unique). TBB at ESR45 brings it down to just over 4 bits. Opera stable is over 11 bits. 4 bits is still significant. Those figures need to be taken with a grain of salt - the database and calculations are based on a lot of old history, and skewed by the nature of the visitors. They do not mimic real world (I think the figures would be a little lower). If you're going to recommend UA string off, then it's quite incorrect to say it's only a small amount of entropy.

IF you disable the UA string, then IMO, you should also make sure that `general.buildID.override` is also disabled otherwise you will stand out, IMO. These settings (and other `general.*.override` prefs) only make sense when the same values are used by a large subset and enforced (and also when all the navigator, <isindex> locale, resource://URIs (yup, we have an extension), and other holes are plugged - which is hopefully what privacy.resistFingerprinting will do with [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1333651) ticket

**Comment by nodiscc** [comment]
Very valid points.

The commented-out UserAgent section could be improved with prefs covering `navigator.platform,oscpu,buildID`.

I see that `general.buildID.override` is already there, was it tested as effective? Also I think that it should be commented out and moved alongside the UserAgent spoofing section, since it makes no sense to spoof one and not the other. @pyllyukko ?

@Thorin-Oakenpants I believe you can reopen this

**Comment by Thorin-Oakenpants** [comment]
I believe you should just use this which I wrote as a complete overhaul (not sure now why I included the language part, excerpt that it is part of the overrides)

```js
/*** 2697: USER AGENT (UA) SPOOFING
     Spoofing your UA to *LOWER* entropy *does* *not* *work*. It may even cause site breakage
     depending on your values. Even if you spoof, like TBB (Tor Browser Bundle) does, as the
     latest ESR, it still *does* *not* *work*. There are two main reasons for this.
       1. Many of the components that make up your UA can be derived by other means. And when
          those values differ, you provide more bits and raise entropy. Examples of leaks include
          navigator objects, resource://URIs, <isindex> locale, feature detection and more.
       2. You are not in a controlled set of significant numbers, where the values are enforced
          by default. It works for TBB because for TBB, the spoofed values ARE their default.
     * We do not recommend UA spoofing yourself, leave it to privacy.resistFingerprinting (see 2699)
     * Values below are for example only based on the current ESR/TBB at the time of writing
***/
/* 2697a: navigator.userAgent leaks in JS
 * [NOTE] setting this will break any UA spoofing add-on whitelisting ***/
   // user_pref("general.useragent.override", "Mozilla/5.0 (Windows NT 6.1; rv:45.0) Gecko/20100101 Firefox/45.0"); // (hidden pref)
/* 2697b: navigator.buildID (see gecko.buildID in about:config) reveals build time
 * down to the second which defeats user agent spoofing and can compromise OS etc
 * [1] https://bugzilla.mozilla.org/show_bug.cgi?id=583181 ***/
   // user_pref("general.buildID.override", "20100101"); // (hidden pref)
/* 2697c: navigator.appName ***/
   //user_pref("general.appname.override", "Netscape"); // (hidden pref)
/* 2697d: navigator.appVersion ***/
   // user_pref("general.appversion.override", "5.0 (Windows)"); // (hidden pref)
/* 2697e: navigator.platform leaks in JS ***/
   // user_pref("general.platform.override", "Win32"); // (hidden pref)
/* 2697f: navigator.oscpu leaks in JS ***/
   // user_pref("general.oscpu.override", "Windows NT 6.1"); // (hidden pref)
/* 2697g: also see 0204 for general.useragent.locale ***/
```

**Comment by Thorin-Oakenpants** [comment]
^^ obviously tweak it to your liking, but group *all* that stuff together

**Comment by pyllyukko** [comment]
> I see that general.buildID.override is already there, was it tested as effective? Also I think that it should be commented out and moved alongside the UserAgent spoofing section, since it makes no sense to spoof one and not the other. @pyllyukko ?

Yes, it was tested. See [here](https://github.com/pyllyukko/user.js/issues/117#issuecomment-184371144). We're leaving it there.

The thing I worry about in this, is that when crafting client-side exploits against the browser, an attacker can practice with the same **exact** version and get to a whole different level of precision to perfect the exploit.

Security is never perfect and as @Thorin-Oakenpants so often points out, that there are still ways to FP the browser, which of course is true. But it's about raising the bar and making the attackers work harder. This applies to privacy as well... even though it doesn't mean squat against advanced FP techniques, it still matters.

Also see the discussion in commit 35b9892933b1a02f1da62224c1fe4ce6e5afced8.

**Comment by Thorin-Oakenpants** [comment]
I can see your point about the buildID giving an exploit the exact information it wants for crafted scripted attacks (security : for anyone reading who doesn't understand, read this article : https://betanews.com/2017/03/22/user-agent-based-attacks-are-a-low-key-risk-that-shouldnt-be-overlooked/ ) and then the override making you part of an extremely small group (tracking/privacy).

But for BOTH these scenarios, these are end of the line measures IMO. There are many layers - common sense, default deny-all JSS/XSS etc. When it comes to FP'ing, there are already lots of other ways still open, combined even more so. With FP we clearly want to jump on anything that makes us unique (such as canavas - this would be high priority and has no security implications). But others have trade-offs.

So the question becomes (and you should always be on an up-to-date browser), do you care about just high the entropy is with the buildID override (I would say you are in an extremely small group), or worry about (a slim to none?) chance of coming across an enabled script that specifically targets your up-to-date browser. Trade-off.

Personally, due to other hardening, this is a line call. Your build ID obfuscation will not prevent zero days, or unpatched known exploits (and you already give away your UA string), and if you are using an up-to-date browser, there's not much more you can do anyway. The chances IMO are so minuscule as to practically be zero, but sites that may FP using your buildID are numerous. Pyllyukko feels the opposite: security trumps privacy in this instance. It also really depends on the user's setup: eg default deny-all JS/XSS. Joe average who uses this and doesn't use uMatrix in default-deny (and/or NS/uBo etc same mode) just opens himself *right up* in FPing terms (very high entropy), but someone like pyllyukko, nodiscc or myself - it's a minor setting with much small repercussions.

I just think that people need to understand this, not lecturing at you guys - this is for the benefit of casual readers I suppose.

Last thought: And this should make the argument for security moot. If you are not UA spoofing, then you are *already* giving away your browser model *and* OS etc, just not the specific build. The buildID becomes immaterial - they already know. If you're on WIndows with FF52 64 so far your chances are two buildIDs (and there were no securities fixes in 52.01). That may be all there ever is until 53 lands. The buildID is useless except when used with UA spoofing.
